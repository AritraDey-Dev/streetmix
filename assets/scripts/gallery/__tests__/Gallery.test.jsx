/* eslint-env jest */
import React from 'react'
import { fireEvent } from '@testing-library/react'
import { renderWithReduxAndIntl } from '../../../../test/helpers/render'
import MOCK_STREET from '../../../../test/fixtures/street.json'
import Gallery from '../Gallery'
import { switchGalleryStreet } from '../view'
import { hideGallery } from '../../store/actions/gallery'

jest.mock('../view')
jest.mock('../../app/errors')
jest.mock('../../streets/thumbnail')
jest.mock('../../streets/xhr')
jest.mock('../../store/actions/gallery', () => ({
  hideGallery: jest.fn(() => ({ type: 'MOCK_ACTION' }))
}))

const initialState = {
  gallery: {
    userId: 'foo',
    visible: true,
    mode: 'GALLERY',
    streets: [MOCK_STREET]
  },
  street: {
    id: '2556be10-df45-11e9-92a0-b5e383de159b'
  },
  user: {
    signedIn: true,
    signInData: {
      userId: 'foo'
    }
  }
}

describe('Gallery', () => {
  it('renders main gallery view for user’s own streets', () => {
    const wrapper = renderWithReduxAndIntl(<Gallery />, { initialState })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders main gallery view for another user’s streets', () => {
    const wrapper = renderWithReduxAndIntl(<Gallery />, {
      initialState: {
        ...initialState,
        user: {
          signedIn: true,
          signInData: {
            userId: 'bar'
          }
        }
      }
    })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders global gallery view', () => {
    const initialState = {
      gallery: {
        visible: true,
        mode: 'GALLERY',
        streets: [MOCK_STREET]
      },
      street: {
        id: '2556be10-df45-11e9-92a0-b5e383de159b'
      },
      user: {
        signedIn: false
      }
    }

    const wrapper = renderWithReduxAndIntl(<Gallery />, { initialState })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders loading', () => {
    const initialState = {
      gallery: {
        visible: true,
        mode: 'LOADING'
      }
    }

    const wrapper = renderWithReduxAndIntl(<Gallery />, { initialState })
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders error', () => {
    const initialState = {
      gallery: {
        visible: true,
        mode: 'ERROR'
      }
    }

    const wrapper = renderWithReduxAndIntl(<Gallery />, { initialState })
    expect(wrapper.getByText('Failed to load the gallery.')).toBeInTheDocument()
  })

  it('closes on shield click', () => {
    const initialState = {
      gallery: {
        visible: true,
        mode: 'GALLERY'
      }
    }

    const wrapper = renderWithReduxAndIntl(<Gallery />, { initialState })
    fireEvent.click(wrapper.container.querySelector('.gallery-shield'))
    expect(hideGallery).toHaveBeenCalledTimes(1)
  })

  describe('street item', () => {
    it('selects street', () => {
      const { getByText } = renderWithReduxAndIntl(<Gallery />, {
        initialState
      })
      fireEvent.click(getByText('Baz'))
      expect(switchGalleryStreet).toHaveBeenCalledWith(initialState.street.id)
    })

    it('deletes street', () => {
      const { getByTitle, queryByTitle } = renderWithReduxAndIntl(<Gallery />, {
        initialState
      })
      fireEvent.click(getByTitle('Delete street'))

      // There's only one street "displayed" so we expect no "delete street"
      // button to be rendered anymore, since that street should be deleted.
      expect(queryByTitle('Delete street')).not.toBeInTheDocument()
    })
  })
})
