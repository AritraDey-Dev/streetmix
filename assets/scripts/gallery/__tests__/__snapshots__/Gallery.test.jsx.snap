// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery renders global gallery view 1`] = `
<DocumentFragment>
  <div
    class="gallery"
  >
    <div
      class="gallery-panel"
    >
      <div
        class="gallery-label"
      >
        All streets
      </div>
      <div
        class="gallery-streets-container"
      >
        <div
          class="streets-scrollable-container"
        >
          <div
            class="streets"
          >
            <div
              class="gallery-street-item"
            >
              <a
                class=""
                href="/foo/435"
              >
                <canvas
                  height="220"
                  width="360"
                />
                <div
                  class="street-name"
                >
                  <div
                    class="street-name-text"
                  >
                    Unnamed St
                  </div>
                </div>
                <span
                  class="gallery-street-item-date"
                >
                  Today at 
                  <time
                    datetime="2019-09-25T19:29:22.547Z"
                    title="2019-09-25T19:29:22.547Z"
                  >
                    3:29 PM
                  </time>
                </span>
                <span
                  class="gallery-street-item-creator"
                >
                  foo
                </span>
              </a>
            </div>
          </div>
          <button
            class="scrollable scroll-left"
            style="left: -15px; opacity: 0; pointer-events: none;"
          >
            <svg
              class="svg-inline--fa fa-chevron-left"
            />
          </button>
          <button
            class="scrollable scroll-right"
            style="right: -15px; opacity: 0; pointer-events: none;"
          >
            <svg
              class="svg-inline--fa fa-chevron-right"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="gallery-shield"
    />
  </div>
</DocumentFragment>
`;

exports[`Gallery renders loading 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="gallery"
  >
    <div
      class="gallery-panel"
    >
      <div
        aria-busy="true"
        aria-live="polite"
        class="gallery-loading"
      >
        <div
          class="loading-spinner"
        />
      </div>
    </div>
    <div
      class="gallery-shield"
    />
  </div>
</DocumentFragment>
`;

exports[`Gallery renders main gallery view for another user’s streets 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="gallery"
  >
    <div
      class="gallery-panel"
    >
      <div
        class="gallery-label"
      >
        <div
          class="avatar"
        >
          <img
            alt="foo"
          />
        </div>
        <div
          class="gallery-user-id"
        >
          foo
        </div>
      </div>
      <div
        class="gallery-street-count"
      >
        1 street
      </div>
      <div
        class="gallery-streets-container"
      >
        <div
          class="streets-scrollable-container"
        >
          <div
            class="streets"
          >
            <div
              class="gallery-street-item"
            >
              <a
                class=""
                href="/foo/435"
              >
                <canvas
                  height="220"
                  width="360"
                />
                <div
                  class="street-name"
                >
                  <div
                    class="street-name-text"
                  >
                    Unnamed St
                  </div>
                </div>
                <span
                  class="gallery-street-item-date"
                >
                  Today at 
                  <time
                    datetime="2019-09-25T19:29:22.547Z"
                    title="2019-09-25T19:29:22.547Z"
                  >
                    3:29 PM
                  </time>
                </span>
              </a>
            </div>
          </div>
          <button
            class="scrollable scroll-left"
            style="left: -15px; opacity: 0; pointer-events: none;"
          >
            <svg
              class="svg-inline--fa fa-chevron-left"
            />
          </button>
          <button
            class="scrollable scroll-right"
            style="right: -15px; opacity: 0; pointer-events: none;"
          >
            <svg
              class="svg-inline--fa fa-chevron-right"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="gallery-shield"
    />
  </div>
</DocumentFragment>
`;

exports[`Gallery renders main gallery view for user’s own streets 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="gallery"
  >
    <div
      class="gallery-panel"
    >
      <div
        class="gallery-label"
      >
        <div
          class="avatar"
        >
          <img
            alt="foo"
          />
        </div>
        <div
          class="gallery-user-id"
        >
          foo
        </div>
      </div>
      <div
        class="gallery-street-count"
      >
        1 street
      </div>
      <div
        class="gallery-streets-container"
      >
        <div
          class="gallery-user-buttons"
        >
          <a
            class="button-like gallery-new-street"
            href="/new"
            target="_blank"
          >
            Create new street
          </a>
          <a
            class="button-like gallery-copy-last-street"
            href="/copy-last"
            target="_blank"
          >
            Make a copy
          </a>
        </div>
        <div
          class="streets-scrollable-container"
        >
          <div
            class="streets"
          >
            <div
              class="gallery-street-item"
            >
              <a
                class=""
                href="/foo/435"
              >
                <canvas
                  height="220"
                  width="360"
                />
                <div
                  class="street-name"
                >
                  <div
                    class="street-name-text"
                  >
                    Unnamed St
                  </div>
                </div>
                <span
                  class="gallery-street-item-date"
                >
                  Today at 
                  <time
                    datetime="2019-09-25T19:29:22.547Z"
                    title="2019-09-25T19:29:22.547Z"
                  >
                    3:29 PM
                  </time>
                </span>
                <button
                  class="gallery-street-item-delete"
                  title="Delete street"
                >
                  <svg
                    class="svg-inline--fa fa-times"
                  />
                </button>
              </a>
            </div>
          </div>
          <button
            class="scrollable scroll-left"
            style="left: -15px; opacity: 0; pointer-events: none;"
          >
            <svg
              class="svg-inline--fa fa-chevron-left"
            />
          </button>
          <button
            class="scrollable scroll-right"
            style="right: -15px; opacity: 0; pointer-events: none;"
          >
            <svg
              class="svg-inline--fa fa-chevron-right"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="gallery-shield"
    />
  </div>
</DocumentFragment>
`;
